Support creation of fat binaries that run on x86_64 and arm64

ahoppen opened the pull request:
Currently, when building an open source toolchain, SourceKit-LSP is only built for x86_64.  Copy the necessary cross-compilation parts from SwiftPM’s build script to also produce a fat sourcekit-lsp executable for both x86_64 and arm64.

rdar://78039145

benlangmuir commented on Utilities/build-script-helper.py:
```diff
@@ -149,7 +181,7 @@ def add_common_args(parser):
     parser.add_argument('--no-clean', action='store_true', help='Don\'t clean the build directory prior to performing the action')
     parser.add_argument('--verbose', '-v', action='store_true', help='enable verbose output')
-    parser.add_argument('--cross-compile-host', help='cross-compile for another host instead')
+    parser.add_argument('--cross-compile-hosts', help='List of cross compile hosts targets', default=[])
```
Is this actually a list, or is it just a string? The default value is `[]`, but I would have expected to see `nargs=*` or similar.

benlangmuir commented on Utilities/build-script-helper.py:
```diff
@@ -51,8 +77,8 @@ def get_swiftpm_options(args):
 
-  if 'ANDROID_DATA' in os.environ or (args.cross_compile_host and re.match(
-    'android-', args.cross_compile_host)):
+  if 'ANDROID_DATA' in os.environ or (args.cross_compile_hosts and re.match(
```
Have you confirmed this line works? Matching a list doesn't seem to work for me.  Or maybe this is only working because it's actually a string right now (see my other comment on the argparse code).

```python
>>> re.match('as', 'asdf')
<re.Match object; span=(0, 2), match='as'>

>>> re.match('as', ['asdf'])
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/AppleInternal/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/re.py", line 191, in match
    return _compile(pattern, flags).match(string)
TypeError: expected string or bytes-like object
```

buttaface commented on Utilities/build-script-helper.py:
```diff
@@ -65,8 +91,12 @@ def get_swiftpm_options(args):
   if args.cross_compile_host:
-    swiftpm_args += ['--destination', args.cross_compile_config]
+    if build_target == 'x86_64-apple-macosx' and args.cross_compile_host == "macosx-arm64":
+        swiftpm_args += ["--arch", "x86_64", "--arch", "arm64"]
```
Would you bring the two Android lines deleted below up here and move the config line deleted two lines above after printing the cross-compilation host? Otherwise, this is going to break [cross-compiling sourcekit-lsp for Android](https://github.com/termux/termux-packages/blob/master/packages/swift/build.sh#L162).

ahoppen commented on Utilities/build-script-helper.py:
```diff
@@ -65,8 +91,14 @@ def get_swiftpm_options(args):
-    swiftpm_args += ['--destination', args.cross_compile_config]
+  build_target = get_build_target(args)
+  cross_compile_hosts = args.cross_compile_hosts
+  if build_target == 'x86_64-apple-macosx' and "macosx-arm64" in cross_compile_hosts:
```
We don’t have any build presets that cross compile from arm64 for x86_64, so I’m not entire sure what to look here (I would be guessing `macosx-x86_64`, but that’s really mostly guessing). So I’d prefer to stick with what SwiftPM does and only support cross-compilation from x86_64 to arm64 for now.

ahoppen commented:
@swift-ci Please test

ahoppen commented:
I simplified the PR a little bit to only take a single `cross-compile-host` (which seems like it was the only supported configuration anyway) and addressed your feedback. It now successfully produced a universal `sourcekit-lsp` binary and I’m happy with the current state. Could you take another look at it?

buttaface commented on Utilities/build-script-helper.py:
```diff
@@ -9,6 +9,32 @@
+    command = [swift_exec, '-print-target-info']
+    if cross_compile:
+      cross_compile_json = json.load(open(args.cross_compile_config))
+      command += ['-target', cross_compile_json["target"]]
```
These three lines and the `cross_compile` flag can be removed as they are unused on macOS. With SPM, [this flag was only set for Android to get the target directory](https://github.com/apple/swift-package-manager/blob/main/Utilities/bootstrap#L221), which is unneeded here.

ahoppen commented:
@swift-ci Please test

benlangmuir commented on Utilities/build-script-helper.py:
```diff
@@ -9,6 +9,29 @@
 import sys
+import json
+
+def error(message):
```
😅 oops, guess we forgot to define this one

benlangmuir commented on Utilities/build-script-helper.py:
```diff
@@ -9,6 +9,29 @@
+import json
+
+def error(message):
+  print("--- %s: error: %s" % (os.path.basename(sys.argv[0]), message))
```
Should this go to stderr?

ahoppen replied:
Good idea. SwiftPM (where I copied the function from) prints to stdout but stderr seems like the better choice.

ahoppen commented:
@swift-ci Please test
